### PR TITLE
Add support for max pool backward

### PR DESCRIPTION
**Pull request by asuhan**

### CODE REVIEW

**Comment by asuhan** [comment]
Build failure is fixed by https://github.com/pytorch/xla/pull/198.

File: torch_xla/csrc/ops/max_pool2d_backward.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    const Value& grad_output, const Value& input,
+    tensorflow::gtl::ArraySlice<const xla::int64> kernel_size,
+    tensorflow::gtl::ArraySlice<const xla::int64> stride,
+    tensorflow::gtl::ArraySlice<const xla::int64> padding) {
+  auto lower_for_shape_fn =
+      [stride, padding,
+       kernel_size](tensorflow::gtl::ArraySlice<const xla::XlaOp> operands)
+      -> xla::XlaOp {
+    XLA_CHECK_EQ(operands.size(), 2)
+        << "Unexpected number of operands: " << operands.size();
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/max_pool2d_backward.cpp]: IMO we can get rid of these extra string messages.
That macros already provides a text like:

"Error: XLA_CHECK_EQ(operands.size(), 2) : 15 vs 2"

Which is a pretty clear interpretation even w/out the extra message.

File: torch_xla/csrc/ops/ops.cpp [hunk 2]
```diff
@@ -151,6 +151,18 @@ NodePtr NllLossBackwardOp(const Value& logits, const Value& labels) {
                             std::move(lower_fn));
 }
 
+NodePtr NotSupportedOp(c10::Symbol node_symbol) {
+  auto lower_fn = [](const ir::Node& node,
+                     ir::LoweringContext* loctx) -> ir::XlaOpVector {
+    XLA_ERROR() << "Node not supported: " << node.ToString();
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/ops.cpp]: XLA_ERROR leads to a noreturn function attribute.

File: torch_xla/csrc/ops/ops.cpp [hunk 3]
```diff
@@ -151,6 +151,18 @@ NodePtr NllLossBackwardOp(const Value& logits, const Value& labels) {
                             std::move(lower_fn));
 }
 
+NodePtr NotSupportedOp(c10::Symbol node_symbol) {
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/ops.cpp]: This op should take a shape.

File: torch_xla/csrc/ops/ops.h [hunk 4]
```diff
@@ -59,6 +59,10 @@ NodePtr NllLossOp(const Value& logits, const Value& labels);
 
 NodePtr NllLossBackwardOp(const Value& logits, const Value& labels);
 
+// Placeholder node which is never to be used. Using it would throw an error
+// during lowering.
+NodePtr NotSupportedOp(c10::Symbol node_symbol);
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/ops.h]: Missing shape.

File: torch_xla/csrc/ops/ops.cpp (hunk 3 above)
Comment (asuhan) [reply torch_xla/csrc/ops/ops.cpp]: Where would it get the shape from? If there's no XLA lowering, we'd have to write a type equation, which is overkill for something we don't want to implement in the first place.

Comment (dlibenzi) [reply torch_xla/csrc/ops/ops.cpp]: It's the same shape as the output, with int64 type AFAICT.
The int64 is a linearized coordinates of the winning index at that position.

File: torch_xla/csrc/ops/max_pool2d_backward.cpp (hunk 1 above)
Comment (asuhan) [reply torch_xla/csrc/ops/max_pool2d_backward.cpp]: Done.

File: torch_xla/csrc/ops/ops.cpp (hunk 2 above)
Comment (asuhan) [reply torch_xla/csrc/ops/ops.cpp]: Done.

File: torch_xla/csrc/ops/ops.h (hunk 4 above)
Comment (asuhan) [reply torch_xla/csrc/ops/ops.h]: Done.

**Pull request merged**